packages/contracts/src/dollar/core/CreditClock.sol

```diff
@@ -0,0 +1,102 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.3;
+
+import "../libs/ABDKMathQuad.sol";
+import "./UbiquityDollarManager.sol";
+
+contract CreditClock {
+    using ABDKMathQuad for uint256;
+    using ABDKMathQuad for bytes16;
+
+    UbiquityDollarManager public manager;
+
+    uint256 public rateStartBlock;
+    bytes16 public rateStartValue;
+    bytes16 public ratePerBlock;
+
+    event SetRatePerBlock(
+        uint256 rateStartBlock, bytes16 rateStartValue, bytes16 ratePerBlock
+    );
+
+    modifier onlyAdmin() {
+        require(
+            manager.hasRole(manager.INCENTIVE_MANAGER_ROLE(), msg.sender),
+            "CreditClock: not admin"
+        );
+        _;
+    }
+
+    /// @param _manager The address of the manager/config contract so we can fetch variables.
+    /// @param _rateStartValue ABDKMathQuad Initial rate.
+    /// @param _ratePerBlock ABDKMathQuad Initial rate change per block.
+    constructor(UbiquityDollarManager _manager, bytes16 _rateStartValue, bytes16 _ratePerBlock) {
+        manager = _manager;
+        rateStartBlock = block.number;
+        rateStartValue = _rateStartValue;
+        ratePerBlock = _ratePerBlock;
+    }
+
+    /// @dev Sets rate to apply from this block onward.
+    /// @param _ratePerBlock ABDKMathQuad New rate per block to apply from this block onward.
+    function setRatePerBlock(bytes16 _ratePerBlock)
+        external
+        onlyAdmin
+    {
+        rateStartValue = calculateRate(rateStartValue, ratePerBlock, block.number - rateStartBlock);
+        rateStartBlock = block.number;
+        ratePerBlock = _ratePerBlock;
+
+        emit SetRatePerBlock(rateStartBlock, rateStartValue, ratePerBlock);
+    }
+
+    /// @dev Calculates b raised to the power of n.
+    /// @param b ABDKMathQuad
+    /// @param n ABDKMathQuad
+    /// @return ABDKMathQuad b ^ n
+    function pow(bytes16 b, bytes16 n)
+        private
+        pure
+        returns (bytes16)
+    {
+        return n.mul(b.log_2()).pow_2();
```

Comment:
for clarity can you add in the comment ` b ^ n == 2^(n*log²(b)` 
I previously used b^n =  e^(n*lnx(b)) but using only binary log and 2^ saves gas :+1: 
the math works the same cos  
2^(n*log²(b)) = (2^(log²(b)))^n = b^n